### PR TITLE
fix gemma4 gradient accumulation loss and last token incorrect labels

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?

Gemma 4 was calculating the CE loss incorrectly and not handling gradient accumulation steps properly, leading to losses scaled up by the value of the gradient accumulation steps rather than letting the built in HF CE loss function handle it correctly with `num_items_in_batch`. Gemma 4's custom loss function was simply using attention mask to filter out tokens, but we already handle this via when `labels` are set to `-100`.

Additionally, when doing multimodal training, the loss often starts off at ~15. This is partly attributed to most multimodal datasets being out of distribution (I tested this by having Gemma 4 generate a response to a prompt and checking the loss for that and it was closs to a more normal ~1 loss. The second reason is that the logit/label shift and could result in the last token in a sequence trying to predict a token like a pad token and it wouldn't be masked out because the original loss function didn't handle this, but using properl `-100` labels does.  

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @Cyrilvallez @zucchini-nlp

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by Cyrilvallez — APPROVED** [review]
Thanks @winglian! Reflected the change in modular, and added gemma3n as well at the same time!
